### ISSUE TITLE
Possible bug in cuda source

### ISSUE THREAD

**Issue by trunda-o** [jdermody/brightwire#22]
Hello again.
I've been getting some suspicious results with convolutional networks and I think I managed to track down the problem into _TensorIm2Col_ method in file _BrightWire.CUDA.Net4.x64/cuda/brightwire.cu_.

I believe on the line 603 https://github.com/jdermody/brightwire/blob/6713e4bb42a4c837376ef1a47212141de747a62b/BrightWire.CUDA.Net4.x64/cuda/brightwire.cu#L603 there should be _yExtent_ instead of _xExtent_ 
and on the line 606 https://github.com/jdermody/brightwire/blob/6713e4bb42a4c837376ef1a47212141de747a62b/BrightWire.CUDA.Net4.x64/cuda/brightwire.cu#L606 there should be _filterHeight_ instead of _filterWidth_.

Could you please take a look on that?

**Comment by jdermody** [comment]
Hi, thanks for that. That code is certainly a bit fiddly!

I refactored the code to pre-calculate the indices and to reference those indices from a buffer. This makes the cuda kernels simpler to understand (and maintain!)

Along the way I noticed some performance improvements and the convolutions are now running much faster. I haven't merged it with master yet but feel free to [take a look at the changes](https://github.com/jdermody/brightwire/tree/cuda_refactor/BrightWire.CUDA.Net4.x64/cuda) if you're interested.
